### DISCUSSION TITLE
How to determine the value of `__mask` at an instance

### DISCUSSION THREAD

**Discussion by lalith-mcw** [ispc/ispc#2840] [Q&A] [answered]
How the compiler determines the value of `__mask` for an instance. It is mentioned that `__mask` acts as a program counter for the instance:
https://ispc.github.io/ispc.html

Are there provisions to modify the value of `__mask`, how to check in low level where the `__mask` is being updated in the program

**Comment by MkazemAkhgary** [comment]
I may be mistaken, but I believe `__mask` is a `varying` read-only value. For each program instance, if the `__mask` value is `-1` (all 1s in binary, like `11111111`), it indicates that the program instance is active; if the value is 0, it indicates that the program instance is inactive.

By the way, `__mask` is not a program counter. It’s an execution mask, and its values are determined at runtime. I don’t think you can modify `__mask` directly—it wouldn’t be safe to do so.

`lanemask()` and `packmask(__mask != 0)` should both return the same value—that is, a bitmask of active program instances. take this with a grain of salt, as this was just my understanding.

**Comment by nurmukhametov** [comment, marked as answer]
### Answer

`__mask` is not a program counter.

> How the compiler determines the value of __mask for an instance.

Compiler determines the `__mask` value considering predicates (control flow conditions).

> how to check in low level where the __mask is being updated in the program

You can be able to see the value of `__mask` in the debugger. I suggest that you can set up the watchpoint for `__mask` to observe moments when it changes.

### Explanation

`__mask` is needed to handle varying control flow. Under control flow statements with `varying` variables,  ISPC provides the illusion of separate control flow for each SIMD lane. The compiler translates if statements into partially predicated instructions. Both branches are executed but a different mask is maintained to indicate active program instances (vector/SIMD lanes). Operations with potential side-effects are conditionally executed based on this mask.

Consider the following example:

```ispc
varying int f(varying int a, varying int b) {
    print("% %\n", a, __mask); // [-2,-1,1,2] [true,true,true,true]
    if (a < 0) {
        print("% %\n", a, __mask); // [-2,-1,((1)),((2))] [true,true,_________,_________]
        a = 0;
        print("% %\n", a, __mask); // [0,0,((1)),((2))] [true,true,_________,_________]
    } else {
        print("% %\n", a, __mask); // [((0)),((0)),1,2] [_________,_________,true,true]
        a += b;
        print("% %\n", a, __mask); // [((0)),((0)),11,12] [_________,_________,true,true]
    }
    print("% %\n", a, __mask); // [0,0,11,12] [true,true,true,true]
    return a;
}

extern "C" uniform int main() {
    varying int x = { -2, -1, 1, 2 };
    varying int y = 10;
    varying int z = f(x, y);
    print("%\n", z);
    return 0;
}
```

Here, assignments to a variable are controlled by an if statement. We suggest that this function is called with all masks on and input `a` values is vector of `[-2, -1, 1, 2]`. On the `true` branch, lanes with values less than `0` are active. Inactive lanes are in braces. Assignment statement changes only values in active lanes, the first two lanes, the last two are left unchanged. On the `false` branch, on the contrary the first two lanes are inactive. Values in the last two lanes are changed. At the merge point, all program instances are active again.

> I may be mistaken, but I believe __mask is a varying read-only value. For each program instance, if the __mask value is -1 (all 1s in binary, like 11111111), it indicates that the program instance is active; if the value is 0, it indicates that the program instance is inactive.

`__mask` is indeed a varying value. Its type is a little bit tricky but it has two possible values `active` and `inactive`. I think it is correct to say that it has a `bool` type with the corresponding `true` and `false`. But I am sure, that the user code must not rely on exact representation, i.e., all 1s or not, it has to be purely internal thing.
